pkg/providers/vsphere/vmlifecycle/bootstrap.go

```diff
@@ -137,57 +140,93 @@ func DoBootstrap(
 		} else {
+			reconfigureNeeded = newHash != vmCtx.VM.Annotations[hashKey]
+			if !reconfigureNeeded {
+				vmCtx.Logger.V(4).Info(
+					"Skipping bootstrap reconfigure as nothing has changed")
+			}
+		}
+
+		if reconfigureNeeded {
 			vmCtx.Logger.V(4).Info("Doing bootstrap reconfigure")
 			if err := doReconfigure(vmCtx, vcVM, configSpec); err != nil {
 				return fmt.Errorf("bootstrap reconfigure failed: %w", err)
 			}
-			if vmCtx.VM.Annotations == nil {
-				vmCtx.VM.Annotations = map[string]string{}
+
+			if newHash != vmCtx.VM.Annotations[hashKey] {
+				vmCtx.Logger.V(4).Info(
+					"Updating bootstrap reconfigure hash", hashKey, newHash)
+				if vmCtx.VM.Annotations == nil {
+					vmCtx.VM.Annotations = map[string]string{}
+				}
+				vmCtx.VM.Annotations[hashKey] = newHash
 			}
-			vmCtx.Logger.V(4).Info(
-				"Updating bootstrap reconfigure hash", hashKey, newHash)
-			vmCtx.VM.Annotations[pkgconst.BootstrapHashConfigSpecAnnotationKey] = newHash
-			return ErrBootstrapReconfigure
+
+			retErr = errors.Join(retErr, ErrBootstrapReconfigure)
 		}
 	}
 
 	if customSpec != nil {
+		const hashKey = pkgconst.BootstrapHashCustomSpecAnnotationKey
+
 		newHash, err := getVimTypeHash(customSpec)
 		if err != nil {
 			return err
 		}
-		hashKey := pkgconst.BootstrapHashCustomSpecAnnotationKey
-		curHash := vmCtx.VM.Annotations[hashKey]
-		if newHash == curHash {
-			vmCtx.Logger.V(4).Info(
-				"Skipping bootstrap customize as nothing has changed")
+
+		var customizeNeeded bool
+		if customizeLatch != nil {
+			customizeNeeded = *customizeLatch
 		} else {
-			vmCtx.Logger.V(4).Info("Doing bootstrap customize")
+			customizeNeeded = newHash != vmCtx.VM.Annotations[hashKey]
+			if !customizeNeeded {
+				vmCtx.Logger.V(4).Info(
+					"Skipping bootstrap customize as nothing has changed")
+			}
+		}
+
+		if customizeNeeded {
+			vmCtx.Logger.V(0).Info("Doing bootstrap customize")
 			if err := doCustomize(vmCtx, vcVM, config, customSpec); err != nil {
 				return fmt.Errorf("bootstrap customize failed: %w", err)
 			}
-			if vmCtx.VM.Annotations == nil {
-				vmCtx.VM.Annotations = map[string]string{}
+
+			if newHash != vmCtx.VM.Annotations[hashKey] {
+				vmCtx.Logger.V(4).Info(
+					"Updating bootstrap customize hash", hashKey, newHash)
+				if vmCtx.VM.Annotations == nil {
+					vmCtx.VM.Annotations = map[string]string{}
+				}
+				vmCtx.VM.Annotations[hashKey] = newHash
 			}
-			vmCtx.Logger.V(4).Info(
-				"Updating bootstrap customize hash", hashKey, newHash)
-			vmCtx.VM.Annotations[pkgconst.BootstrapHashCustomSpecAnnotationKey] = newHash
-			return ErrBootstrapCustomize
+
+			retErr = errors.Join(retErr, ErrBootstrapCustomize)
 		}
 	}
 
-	return nil
+	if customizeLatch != nil {
+		// Set latch to false so that a later customization on has to be
+		// explicitly requested.
+		vmCtx.Logger.V(4).Info(
+			"Setting customization latch to false after successful customization")
+		*customizeLatch = false
+	}
```

Comment:
Check `customizeLatch` value is actually true here, so that we don't log `successful customization` when there's no customization at all?